### PR TITLE
feat(es/minifier): Make name mangler understand block scoping

### LINKED ISSUE #5766: Name mangler of minifier should understand block scoping
While working on https://github.com/swc-project/swc/pull/5763, I found that terser reuses variable names in more cases compared to swc.

## Input

```js
function test(foo, bar) {
    try {
        const bar = {};
        throw "PASS";
    } catch (error) {
        return bar(error);
    }
}
console.log(test(null, (x) => x));

```

## Current

```js

function n(n, o) {
    try {
        const t = {};
        throw "PASS";
    } catch (c) {
        return o(c);
    }
}
console.log(n(null, (n)=>n));

```

## terser output

```js

function n(n, o) {
    try {
        const n = {};
        throw "PASS";
    } catch (n) {
        return o(n);
    }
}
console.log(n(null, (n) => n));

```

### PR DESCRIPTION
**Pull request by iheyunfei**
**Description:**

**Related issue:**

 - Closes https://github.com/swc-project/swc/issues/5090.
 - Closes https://github.com/swc-project/swc/issues/5766.

### CODE REVIEW

File: crates/swc/tests/tsc-references/objectRestForOf.2.minified.js [hunk 1]
```diff
@@ -4,14 +4,14 @@ import _object_spread from "@swc/helpers/src/_object_spread.mjs";
 import _object_spread_props from "@swc/helpers/src/_object_spread_props.mjs";
 import _object_without_properties from "@swc/helpers/src/_object_without_properties.mjs";
 for (let _ref of array){
-    var _ref1, { x  } = _ref;
+    var _ref, { x  } = _ref;
```
Comment (kdy1) [inline crates/swc/tests/tsc-references/objectRestForOf.2.minified.js:7]: I think this can be a problem

Comment (iheyunfei) [reply crates/swc/tests/tsc-references/objectRestForOf.2.minified.js]: Fixed.

File: crates/swc/tests/tsc-references/for-of55.1.normal.js [hunk 2]
```diff
@@ -2,6 +2,6 @@
 let v = [
     1
 ];
-for (let v1 of v1){
-    v1;
+for (let v of v){
```
Comment (kdy1) [inline crates/swc/tests/tsc-references/for-of55.1.normal.js:5]: This is wrong

Comment (iheyunfei) [reply crates/swc/tests/tsc-references/for-of55.1.normal.js:5]: I'm not sure. I look into the input file https://github.com/swc-project/swc/blob/main/crates/swc_ecma_parser/tests/tsc/for-of55.ts.

Is the original input wrong already?

Comment (kdy1) [reply crates/swc/tests/tsc-references/for-of55.1.normal.js:5]: I see. My bad. Thank you!

**Review by kdy1 — APPROVED** [review]
Thank you!

---

swc-bump:
 - swc_ecma_transforms_base

**Review by jridgewell — APPROVED** [review]

**Pull request merged**